src/data/list/defs.lean

```diff
@@ -457,6 +468,31 @@ def permutations_aux : list α → list α → list (list α) :=
 def permutations (l : list α) : list (list α) :=
 l :: permutations_aux l []
 
+/-- `permutations'_aux t ts` inserts `t` into every position in `ts`, including the last.
+This function is intended for use in specifications, so it is simpler than `permutations_aux2`,
+which plays roughly the same role in `permutations`.
+
+Note that `(permutations_aux2 t [] [] ts id).2` is similar to this function, but skips the last
+position:
+
+    permutations'_aux 10 [1, 2, 3] =
+      [[10, 1, 2, 3], [1, 10, 2, 3], [1, 2, 10, 3], [1, 2, 3, 10]]
+    (permutations_aux2 10 [] [] [1, 2, 3] id).2 =
+      [[10, 1, 2, 3], [1, 10, 2, 3], [1, 2, 10, 3]] -/
+@[simp] def permutations'_aux (t : α) : list α → list (list α)
+| []      := [[t]]
+| (y::ys) := (t :: y :: ys) :: (permutations'_aux ys).map (cons y)
+
+/-- List of all permutations of `l`. This version of `permutations` is less efficient but has
+simpler definitional equations. The permutations are in a different order.
```

Comment:
I think it's worth mentioning the lemma that shows their equivalence, since it lives in a different file.
```suggestion
simpler definitional equations. The permutations are in a different order, but are equal up to permutation,
as shown by `list.permutations_perm_permutations'`
```